File: 04_quicksort/golang/04_recursive_max.go
package main

import "fmt"

func main() {
	fmt.Println(max([]int{1, 5, 10, 25, 16, 1}))
}

func max(arr []int) int {

	if len(arr) == 1 {
		return arr[0]
	}

	if arr[0] > max(arr[1:]) {
		return arr[0]
	}
	return max(arr[1:])
}

File: 04_quicksort/ts/04_recursive_max.ts
function maxSum(arr: number[]): number | null {
    if (arr.length === 0) {
        return null;
    }

    if (arr.length === 1) {
        return arr[0];
    }

    const subMax = maxSum(arr.slice(1));

    return arr[0] > subMax ? arr[0] : subMax;
};

console.log(maxSum([1, 2, 4, 3]));